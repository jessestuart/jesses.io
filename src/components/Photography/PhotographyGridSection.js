import React from 'react'
import ImageZoom from 'react-medium-image-zoom'
import _ from 'lodash'
import { DateTime } from 'luxon'

import {
  ImageZoomGrid,
  ImageZoomGridElement,
  PhotographySectionHeader,
} from '.'

const PhotographyGridSection = ({ datetime, images, slug }: Props) =>
  _.isNil(images) ? null : (
    <section className="center flex flex-column w-75 justify-center pv5">
      <PhotographySectionHeader datetime={datetime} href={slug} />
      <ImageZoomGrid>
        {images.map(image => (
          <ImageZoomGridElement key={image.src} aspectRatio={image.aspectRatio}>
            <ImageZoom image={{ src: image.src }} />
          </ImageZoomGridElement>
        ))}
      </ImageZoomGrid>
    </section>
  )

type Props = {
  datetime: DateTime,
  images: Array,
  slug: String,
}

export default PhotographyGridSection
