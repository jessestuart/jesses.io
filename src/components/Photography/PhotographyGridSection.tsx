import 'react-image-lightbox/style.css'

import _ from 'lodash'
import { DateTime } from 'luxon'
import React, { Children, SyntheticEvent, useState } from 'react'
import { Flex } from 'rebass/styled-components'

import Carousel, { Modal, ModalGateway } from '@jesses/react-images'

import { useDimensions, useMedia } from 'utils/hooks'
import {
  ImageZoomGridElement,
  PhotographySectionHeader,
  SeeMoreLink,
} from 'components/Photography'

interface Props {
  datetime: DateTime
  // The images to be *currently* displayed for this section.
  images: any[]
  slug?: string
  totalNumImages?: number
}

const PhotographyGridSection = (props: Props) => {
  const { datetime, images, slug = '/#', totalNumImages = 0 } = props
  const [isLightboxOpen, setIsLightboxOpen] = useState(false)
  const [index, setLightboxIndex] = useState(0)

  const sortedImages = _.sortBy(images, 'EXIF.DateTimeOriginal')
  const reactImages: any[] = _.map(sortedImages, 'childImageSharp.sizes')

  const indexIsNumber = (
    selectedIndex: number | SyntheticEvent | undefined,
  ): selectedIndex is number => {
    return selectedIndex != null && typeof selectedIndex == 'number'
  }
  const toggleModal = (selectedIndex?: number | SyntheticEvent | undefined) => {
    if (indexIsNumber(selectedIndex)) {
      setLightboxIndex(selectedIndex)
    }
    setIsLightboxOpen(!isLightboxOpen)
  }

  // Tie media queries to the number of columns.
  const columns = useMedia(
    ['(min-width: 1536px)', '(min-width: 1024px)', '(min-width: 512px)'],
    [4, 3, 2],
    1,
  )
  // Measure the width of the container element.
  // @ts-ignore
  const [ref, { width }] = useDimensions()

  // Form a grid of stacked images, using the width & column calculations we got
  // from the `useMedia` and `useDimensions` hooks above.
  const heights: number[] = new Array(columns).fill(0) // Each column gets a height starting with zero
  const gridItems = sortedImages.map((child, index) => {
    const { childImageSharp } = child
    const aspectRatio = _.get(childImageSharp, 'sizes.aspectRatio')
    const imageWidth = child.width || width / columns
    const height = _.isFinite(child.height)
      ? child.height
      : imageWidth * (2 / aspectRatio)
    // Masonry-grid placing â€” positions each tile sequentially into the
    // smallest column available.
    const column = heights.indexOf(Math.min(...heights)) || 0

    const x = (width / columns) * column
    const y = (heights[column] += height / 2) - height / 2
    return {
      ...child,
      columns,
      height: height / 2,
      index,
      width: width / columns,
      x,
      y,
    }
  })

  return (
    <>
      <PhotographySectionHeader datetime={datetime} href={slug} />
      <Flex
        ref={ref as any}
        className="mt2 mb4"
        // Define height in `style` to avoid creating a new className on each
        // resize / render.
        sx={{ height: Math.max(...heights), position: 'relative' }}
      >
        {width != null &&
          gridItems.map((item, key) => {
            const { height, width, x, y } = item
            return (
              <div
                key={key}
                style={{
                  height,
                  left: x,
                  position: 'absolute',
                  top: y,
                  width,
                }}
              >
                <ImageZoomGridElement
                  image={item}
                  onClick={() => toggleModal(key)}
                />
              </div>
            )
          })}
      </Flex>
      <SeeMoreLink totalNumImages={totalNumImages} href={slug} />
      <ModalGateway>
        {isLightboxOpen && _.size(reactImages) > 0 ? (
          <Modal
            allowFullscreen={false}
            onClose={toggleModal}
            closeOnBackdropClick={true}
          >
            <Carousel
              currentIndex={index}
              styles={{
                header: base => ({
                  ...base,
                  zIndex: -999,
                }),
              }}
              views={reactImages}
            />
          </Modal>
        ) : null}
      </ModalGateway>
    </>
  )
}

export default PhotographyGridSection
