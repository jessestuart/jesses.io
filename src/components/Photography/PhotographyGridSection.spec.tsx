import PhotographyGridSection from 'components/Photography/PhotographyGridSection'
import { DateTime } from 'luxon'
import React from 'react'
import renderer from 'react-test-renderer'

import image2 from '../../../test/images/DSC01983.jpg'
import image1 from '../../../test/images/DSC02630.jpg'

const datetime = DateTime.utc(2019, 5, 1, 1, 1, 1)

test('Render PhotographyGridSection component.', () => {
  const tree = renderer
    .create(
      <PhotographyGridSection datetime={datetime} images={[]} imageCount={0} />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('Render PhotographyGridSection component w/ images.', () => {
  const tree = renderer
    .create(
      <PhotographyGridSection
        datetime={datetime}
        imageCount={2}
        images={[image1, image2]}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
