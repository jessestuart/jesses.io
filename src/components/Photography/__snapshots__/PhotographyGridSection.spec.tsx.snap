// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render PhotographyGridSection component w/ images. 1`] = `
.c1 {
  border-bottom: 2px solid rgba(255,58,92,0.8);
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1::before {
  background-color: rgba(148,117,204,0.8);
  bottom: -2px;
  content: '';
  height: 2px;
  left: 0;
  position: absolute;
  -webkit-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  -webkit-transition: all 0.25s ease-in-out 0s;
  transition: all 0.25s ease-in-out 0s;
  visibility: hidden;
  width: 100%;
}

.c1:hover:before {
  -webkit-transform: scaleX(1);
  -ms-transform: scaleX(1);
  transform: scaleX(1);
  visibility: visible;
}

.c2 {
  position: relative;
  display: grid;
  grid-gap: 1em;
  grid-auto-rows: minmax(4em,auto);
}

.c0 {
  background-color: #fbfafc;
  box-shadow: rgba(25,17,34,0.1) 0px 3px 10px;
}

@media (min-width:60em) {
  .c2 {
    grid-template-columns: repeat(auto-fill,minmax(20vw,1fr));
  }
}

@media (max-width:60em) and (min-width:45em) {
  .c2 {
    grid-template-columns: repeat(auto-fill,minmax(18em,1fr));
  }
}

@media (max-width:45em) {
  .c2 {
    grid-template-columns: repeat(auto-fill,minmax(50vw,1fr));
  }
}

@media (max-width:45em) {
  .c0 {
    margin: 1rem !important;
    padding: 1rem 2rem !important;
  }
}

<section
  className="c0 br2 center mv5-ns mw7 mw9-ns pa5-ns"
  style={
    Object {
      "transition": "all 0.5s",
    }
  }
>
  <a
    className="c1 alegreya-sans db f2 fw5 header-primary mb3 tr w-100"
    href="/true/#"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    01 May 2019
  </a>
  <div
    className="c2"
  />
</section>
`;

exports[`Render PhotographyGridSection component. 1`] = `null`;
